### PR TITLE
Add overload_classmethod

### PR DESCRIPTION
**Pull request by sklam**
- Add `overload_classmethod`
- Add `typeof()` on subclasses of `numba.core.types.Type`

### CODE REVIEW

File: numba/typed/typeddict.py [hunk 1]
```diff
@@ -209,7 +209,7 @@ def copy(self):
 
 
 # XXX: should we have a better way to classmethod
```
Comment (stuartarchibald) [inline numba/typed/typeddict.py:211]:
```suggestion
```
we do, this PR!? :)

File: numba/core/extending.py [hunk 2]
```diff
@@ -217,19 +235,18 @@ def take_impl(arr, indices):
...
 
-    return decorate
+def overload_classmethod(typ, attr, **kwargs):
+    """
+    A decorator marking the decorated function as typing and implementing
+    classmethod *attr* for the given Numba type in nopython mode.
+
+
+    Similar to ``overload_method``.
+    """
```
Comment (stuartarchibald) [inline numba/core/extending.py:248]: Should add this (and an example) to the high-level extending docs?

File: numba/typed/typedlist.py [hunk 3]
```diff
@@ -444,7 +444,7 @@ def __repr__(self):
 
 
 # XXX: should we have a better way to classmethod
-@overload_method(TypeRef, 'empty_list')
+@overload_classmethod(ListType, 'empty_list')
```
Comment (stuartarchibald) [inline numba/typed/typedlist.py:447]: I guess this means `List.empty()` can now written as an alias of this?

File: numba/typed/typedlist.py [hunk 4]
```diff
@@ -444,7 +444,7 @@ def __repr__(self):
 
 
 # XXX: should we have a better way to classmethod
```
Comment (stuartarchibald) [inline numba/typed/typedlist.py:446]:
```suggestion
```

File: numba/core/typing/typeof.py [hunk 5]
```diff
@@ -94,6 +94,9 @@ def _typeof_type(val, c):
     if issubclass(val, np.generic):
         return types.NumberClass(numpy_support.from_dtype(val))
 
+    if issubclass(val, types.Type):
+        return types.TypeRef(val)
```
Comment (stuartarchibald) [inline numba/core/typing/typeof.py:98]: It'd might be useful to make the repr on `TypeRef` more verbose? Using `class MyArray(types.Array): pass`, 
this:
```python
print(typeof(MyArray(types.float64, 3, 'C', True)))
```
gives
```
typeref[readonly array(float64, 3d, C)]
```
Maybe not for this PR though?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, few small things to resolve else looks good.

File: numba/typed/typedlist.py (hunk 3 above)
Comment (sklam) [reply numba/typed/typedlist.py:447]: Yes, we can finally unify the spelling for Dict and List.

File: numba/core/typing/typeof.py (hunk 5 above)
Comment (sklam) [reply numba/core/typing/typeof.py:98]: i made an issue to track this: https://github.com/numba/numba/issues/7055

File: numba/core/extending.py (hunk 2 above)
Comment (sklam) [reply numba/core/extending.py]: Added docs and coverted `overload_method` and `overload_classmethod` to use sphinx autofunction in the high-level extending docs.

File: numba/typed/typeddict.py (hunk 1 above)
Comment (sklam) [reply numba/typed/typeddict.py]: Yes finally

File: numba/core/typing/typeof.py (hunk 5 above)
Comment (stuartarchibald) [reply numba/core/typing/typeof.py:98]: thanks

File: numba/typed/typedlist.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/typed/typedlist.py:446]: Great!

File: numba/core/extending.py [hunk 6]
```diff
@@ -245,6 +245,22 @@ def overload_classmethod(typ, attr, **kwargs):
 
 
     Similar to ``overload_method``.
+
+
+    Here is an example implementing a classmethod on array types to call
```
Comment (stuartarchibald) [inline numba/core/extending.py:250]:
```suggestion
    Here is an example implementing a classmethod the Array type to call
```

File: numba/core/extending.py [hunk 7]
```diff
@@ -217,19 +235,34 @@ def take_impl(arr, indices):
...
+    ``np.arange()``::
+
+        @overload_classmethod(types.Array, "make")
+        def ov_make(cls, nitems):
+            def impl(cls, nitems):
+                return np.arange(nitems)
+            return impl
+
+    The above code will allow the following code to work in jit-compiled code::
```
Comment (stuartarchibald) [inline numba/core/extending.py:259]:
```suggestion
    The above code will allow the following to work in jit-compiled code::
```

File: numba/core/extending.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/core/extending.py]: Thanks, I've checked the rendering locally.

File: numba/core/extending.py (hunk 6 above)
Comment (sklam) [reply numba/core/extending.py:250]:
```suggestion
    Here is an example implementing a classmethod on the Array type to call
```

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes. Great to see this implemented as part of the `numba.extending` API.

**Pull request merged**